### ISSUE TITLE
System.Text.Json source generator doesn't work with native AOT `IlcDisableReflection`

### ISSUE THREAD

**Issue by cfbao** [dotnet/runtime#68093]
### Description

System.Text.Json source generator generates code (that calls code) that uses reflection, hence incompatible with the `IlcDisableReflection` native AOT option.

Was expecting that source generator can eliminate reflection and work with `IlcDisableReflection`.

offending line seems to be this:
https://github.com/dotnet/runtime/blob/736883624a7ef175729cf082e12ce18f236d8524/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs#L22

### Reproduction Steps

* Follow [native AOT sample app's readme](https://github.com/dotnet/samples/blob/main/core/nativeaot/HelloWorld/README.md) to create the project
* replace `Program.cs` with the following
```csharp
using System.Text.Json;
using System.Text.Json.Serialization;

MyClass stuff = new() { Prop = "value" };
string json = JsonSerializer.Serialize( stuff, MyJsonContext.Default.MyClass );
Console.WriteLine( json );

internal record MyClass {
	public string? Prop { get; set; }
}

[JsonSerializable( typeof( MyClass ) )]
internal partial class MyJsonContext : JsonSerializerContext { }
```
* add `<IlcDisableReflection>true</IlcDisableReflection>` to csproj file
* run `dotnet publish -r <RID>`
* run the produced exe

### Expected behavior

program prints `{"Prop":"value"}` on the console

### Actual behavior

the following error shows up
```
Unhandled Exception: EETypeRva:0x002EFD88: Reflection_Disabled
   at Internal.Reflection.RuntimeTypeInfo.get_Assembly() + 0x59
   at System.Text.Json.Serialization.JsonConverter`1..ctor() + 0xc8
   at System.Text.Json.Serialization.JsonResumableConverter`1..ctor() + 0x18
   at System.Text.Json.Serialization.Converters.JsonMetadataServicesConverter`1..ctor(Func`1, ConverterStrategy) + 0x30
   at System.Text.Json.Serialization.Metadata.SourceGenJsonTypeInfo`1.GetConverter(JsonObjectInfoValues`1) + 0x194
   at System.Text.Json.Serialization.Metadata.SourceGenJsonTypeInfo`1..ctor(JsonSerializerOptions, JsonObjectInfoValues`1) + 0x78
   at System.Text.Json.Serialization.Metadata.JsonMetadataServices.CreateObjectInfo[T](JsonSerializerOptions, JsonObjectInfoValues`1) + 0x67
   at MyJsonContext.get_Stuff() + 0x2cd
   at Program.<Main>$(String[]) + 0x73
```

### Configuration

.NET SDK 7.0.100-preview.3.22179.4, commit c48f2c30ee
Windows 10.0.18363, RID: win10-x64
Host version: 7.0.0-preview.3.22175.4, commit 162f83657c

**Comment by jkotas** [comment]
`IlcDisableReflection` mode is very restrictive. It disables majority of the reflection APIs, even the ones that are generally ok with AOT. We do not plan support it in its current form.

https://github.com/dotnet/runtime/issues/67193
https://github.com/dotnet/runtimelab/issues/635#issuecomment-1013898231

**Comment by layomia** [comment]
Current work to support `JsonSerialzier`'s conformance with AOT analysis: https://github.com/dotnet/runtime/pull/68464, https://github.com/dotnet/runtime/issues/68878.

**Comment by eiriktsarpalis** [comment]
I think we can close this as wontfix. Most of the System.Text.Json.Serialization infrastructure depends on some degree of reflection, even in source generated scenaria. We likely won't be able to change this in the future.

That being said, fast-path serialization in source gen should not require any reflection in principle, however fast-path deserialization is not yet implemented (#55043) and it is still missing features compared to the metadata-based serializer.
